tripper/dataset/dataset.py

```diff
@@ -0,0 +1,942 @@
+
+"""
+
+# pylint: disable=invalid-name,redefined-builtin,import-outside-toplevel
+import functools
+import io
+import json
+import re
+import secrets  # From Python 3.9 we could use random.randbytes(16).hex()
+import warnings
+from pathlib import Path
+from typing import TYPE_CHECKING
+from urllib.parse import urlparse
+
+import requests
+import yaml  # type: ignore
+
+from tripper import DCAT, OTEIO, RDF, Triplestore
+from tripper.utils import AttrDict, as_python
+
+if TYPE_CHECKING:  # pragma: no cover
+    from typing import Any, Iterable, List, Mapping, Optional, Sequence, Union
+
+    from tripper.utils import Triple
+
+# Cache decorator
+cache = (
+    functools.cache  # new in Python 3.9, smaller and faster than lru_cache()
+    if hasattr(functools, "cache")
+    else functools.lru_cache(maxsize=1)
+)
+
+
+# Pytest can't cope with this
+# EMMO = Namespace(
+#     iri="https://w3id.org/emmo#",
+#     label_annotations=True,
+#     check=True,
+# )
+
+CONTEXT_PATH = (
+    Path(__file__).parent.parent / "context" / "0.2" / "context.json"
+).as_uri()
+
+# __TODO__: Update URI when merged to master
+CONTEXT_URL = (
+    "https://raw.githubusercontent.com/EMMC-ASBL/tripper/refs/heads/"
+    "dataset/tripper/context/0.2/context.json"
+)
+
+MATCH_PREFIXED_IRI = re.compile(
+    r"^([a-z0-9]*):([a-zA-Z_]([a-zA-Z0-9_/+-]*[a-zA-Z0-9_+-])?)$"
+)
+
+dicttypes = {
+    "parser": {
+        "datadoc_label": "parsers",
+        "@type": OTEIO.Parser,
+    },
+    "generator": {
+        "datadoc_label": "generators",
+        "@type": OTEIO.Generator,
+    },
+    "accessService": {
+        "datadoc_label": "dataServices",
+        "@type": DCAT.DataService,
+    },
+    "distribution": {
+        "datadoc_label": "distributions",
+        "@type": DCAT.Distribution,
+    },
+    "dataset": {
+        "datadoc_label": "datasets",
+        "@type": DCAT.Dataset,
+        # "https://w3id.org/emmo#EMMO_194e367c_9783_4bf5_96d0_9ad597d48d9a",
+    },
+}
+
+
+def save(
+    ts: Triplestore,
+    data: bytes,
+    class_iri: "Optional[str]" = None,
+    dataset: "Optional[Union[str, dict]]" = None,
+    distribution: "Optional[Union[str, dict]]" = None,
+    generator: "Optional[str]" = None,
+    prefixes: "Optional[dict]" = None,
+    use_sparql: "Optional[bool]" = None,
+) -> None:
+    """Saves a documented dataset to a data resource.
+
+    Arguments:
+        ts: Triplestore to load data from.
+        data: Bytes representation of the dataset to save.
+        class_iri: IRI of a class in the ontology (e.g. an `emmo:DataSet`
+            subclass) that describes the dataset that is saved.
+            Is used to select the `distribution` if that is not given.
+            If `distribution` is also given, a
+            `dcat:distribution value <distribution>` restriction will be
+            added to `class_iri`
+        dataset: IRI of dataset for the data to be saved.
+            Or a dict with additional documentation of the dataset.
```

Comment:
This is confusing. What is the IRI. Does it already exist? Or is this something we have to invent/set here.

Also why are we allowed to set an IRI or give additional information, should it not be possible to give both simultaneously?